llvm/lib/MC/MCCASObjectV1.cpp

```diff
@@ -936,13 +937,83 @@ Error MCCASBuilder::createStringSection(
   return Error::success();
 }
 
+/// Reads and returns the length field of a debug info header contained in
+/// Reader, assuming Reader is positioned at the beginning of the section.
+/// The Reader's state is advanced to the first byte after the section.
+Expected<size_t> getSizeFromDebugHeaderAndSkip(BinaryStreamReader &Reader) {
+  // From Dwarf 5 section 7.4:
+  // In the 32-bit DWARF format, an initial length field [...] is an unsigned
+  // 4-byte integer (which must be less than 0xfffffff0);
+  uint32_t Word1;
+  if (auto E = Reader.readInteger(Word1))
+    return E;
+
+  // TODO: handle 64-bit DWARF format.
+  assert(Word1 < 0xfffffff0);
+
+  if (auto E = Reader.skip(Word1))
+    return E;
+  return Word1;
+}
+
+/// For each compile unit inside the __debug_info section, create a
+/// corresponding DebugInfoRef CAS object.
+Error MCCASBuilder::splitDebugInfoCUs() {
+  auto *DebugInfoSection =
+      Asm.getContext().getObjectFileInfo()->getDwarfInfoSection();
+  if (!DebugInfoSection)
+    return Error::success();
+
+  const auto &FragmentList = DebugInfoSection->getFragmentList();
+  assert(FragmentList.size() == 1 && "One fragment in debug str section");
+
+  startSection(DebugInfoSection);
+
+  // Store relocation information into the MCCASBuilder's state.
+  const auto &DebugFragment = cast<MCDataFragment>(FragmentList.front());
+  auto RelocIt = RelMap.find(&DebugFragment);
+  if (RelocIt != RelMap.end())
+    SectionRelocs.append(RelocIt->second.begin(), RelocIt->second.end());
+
+  ArrayRef<char> DebugInfoData = DebugFragment.getContents();
```

Comment:
This doesn't make sense to me, shouldn't you be iterating over all the fragments in the section? The debug info section can have multiple fragments and we want to merge them into one giant ArrayRef and then split it up.